docs/mutation-testing.md

```diff
@@ -0,0 +1,109 @@
+   ```sh
+    awk '
+        /^diff --git/ {
+            diff_line = $0
+            getline
+            if ($0 !~ /^(deleted file mode|similarity index)/) {
+                print diff_line
+                print
+            }
+        }
+        !/^(diff --git|deleted file mode|similarity index|rename from|rename to)/ {print}
+    ' git.diff > processed.diff
+    
+    # Extract mutants based on the processed diff
+    cargo mutants --in-diff processed.diff --list > all_mutants.txt
+
+    # Create a directory for organizing mutants
+    mkdir -p mutants_by_package
+    
+    # Organize mutants into files based on their main folder
+    while IFS= read -r line; do
+          package=$(echo "$line" | cut -d'/' -f1)
+
+          case $package in
+            "stackslib")
+              echo "$line" >> "mutants_by_package/stackslib.txt"
+              ;;
+            "testnet")
+              echo "$line" >> "mutants_by_package/stacks-node.txt"
+              ;;
+            "stacks-signer")
+              echo "$line" >> "mutants_by_package/stacks-signer.txt"
+              ;;
+            *)
+              echo "$line" >> "mutants_by_package/small-packages.txt"
+              ;;
+          esac
+        done < all_mutants.txt
+   ```
+4. based on the package required to run the mutants for  
+   a. stackslib package 
+   ```sh
+    regex_pattern=$(sed 's/[][()\.^$*+?{}|]/\\&/g' "mutants_by_package/stackslib.txt" | paste -sd'|' -)
+
+    cargo mutants --timeout-multiplier 1.5 --no-shuffle -vV \
+        -F "$regex_pattern" \
+        -E ": replace .{1,2} with .{1,2} in " \
+        --output "./stackslib_mutants" \
+        --test-tool=nextest \
+        -- --all-targets --test-threads 1
+   ```
+   b. stacks-node (testnet) package
+   ```sh
+    regex_pattern=$(sed 's/[][()\.^$*+?{}|]/\\&/g' "mutants_by_package/testnet.txt" | paste -sd'|' -)
+
+    cargo mutants --timeout-multiplier 1.5 --no-shuffle -vV \
+        -F "$regex_pattern" \
+        -E ": replace .{1,2} with .{1,2} in " \
+        --output "./testnet_mutants" \
+        --test-tool=nextest \
+        -- --all-targets --test-threads 1
+   ```
+   c. stacks-signer
+   ```sh
+    regex_pattern=$(sed 's/[][()\.^$*+?{}|]/\\&/g' "mutants_by_package/stacks-signer.txt" | paste -sd'|' -)
+
+    cargo mutants --timeout-multiplier 1.5 --no-shuffle -vV \
+        -F "$regex_pattern" \
+        -E ": replace .{1,2} with .{1,2} in " \
+        --output "./stacks-signer_mutants" \
+        --test-tool=nextest \
+        -- --all-targets --test-threads 1
+   ```
+   d. all other packages combined  
+    ```sh
+    regex_pattern=$(sed 's/[][()\.^$*+?{}|]/\\&/g' "mutants_by_package/small-packages.txt" | paste -sd'|' -)
+
+    cargo mutants --timeout-multiplier 1.5 --no-shuffle -vV \
+        -F "$regex_pattern" \
+        -E ": replace .{1,2} with .{1,2} in " \
+        --output "./small-packages_mutants" \
+        --test-tool=nextest \
+        -- --all-targets --test-threads 1
+   ```
```

Comment:
what should the results be? can you share an example what the output should look like, as well as any text about how long you may expect to run this test for?

Reply:
The mutants have a default display of expected time remaining and the time already passed. It displays it after running the initial test suite without mutations. Besides that display, I cannot display expected times as they are related to the capabilities of the machine they are run on.

During the execution, it will display the results for the mutants runs like this
```
caught   stacks-common/src/deps_common/bitcoin/blockdata/script.rs:198:5: replace build_scriptint -> Vec<u8> with vec![] in 16.2s build + 45.6s test
caught   stacks-common/src/deps_common/bitcoin/blockdata/script.rs:198:5: replace build_scriptint -> Vec<u8> with vec![0] in 2.0s build + 45.7s test
caught   stacks-common/src/deps_common/bitcoin/blockdata/script.rs:198:5: replace build_scriptint -> Vec<u8> with vec![1] in 2.0s build + 45.6s test
 WARN nextest process exited with unexpected code (not TEST_RUN_FAILED) code=101
unviable clarity/src/vm/analysis/type_checker/v2_05/natives/mod.rs:552:9: replace TypedNativeFunction::type_native_function -> Result<TypedNativeFunction, CheckErrors> with Ok(Default::default()) in 3.7s build
 WARN nextest process exited with unexpected code (not TEST_RUN_FAILED) code=101
unviable clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs:339:5: replace check_special_equals -> TypeResult with Default::default() in 3.1s build
 WARN nextest process exited with unexpected code (not TEST_RUN_FAILED) code=101
unviable clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs:725:9: replace TypedNativeFunction::type_native_function -> Result<TypedNativeFunction, CheckErrors> with Ok(Default::default()) in 3.1s build
 WARN nextest process exited with unexpected code (not TEST_RUN_FAILED) code=101
unviable clarity/src/vm/analysis/type_checker/v2_1/natives/sequences.rs:61:5: replace check_special_map -> TypeResult with Default::default() in 3.3s build
```

and at the end it will display something like this

```
66 mutants tested in 20m 48s: 13 missed, 19 caught, 34 unviable
```

I'm open to add any changes to it.

Reply:
it think that would be fine - just something to compare with

Reply:
I'll mock it locally to know how the output should look like if more packages are mutated to create a modular comparison and then add it.

Reply:
```
./contrib/tools/local-mutation-testing.sh
     Ignored package `cargo-mutants v24.7.1` is already installed, use --force to override
Running mutants for stacks-signer (       3 mutants)
Found 3 mutants to test
ok       Unmutated baseline in 114.8s build + 9.4s test
 INFO Auto-set test timeout to 20s
MISSED   stacks-signer/src/runloop.rs:424:9: replace <impl SignerRunLoop for RunLoop<Signer, T>>::run_one_pass -> Option<Vec<SignerResult>> with None in 3.0s build + 9.3s test
MISSED   stacks-signer/src/runloop.rs:424:9: replace <impl SignerRunLoop for RunLoop<Signer, T>>::run_one_pass -> Option<Vec<SignerResult>> with Some(vec![]) in 2.3s build + 8.9s test
 WARN nextest process exited with unexpected code (not TEST_RUN_FAILED) code=101
unviable stacks-signer/src/runloop.rs:424:9: replace <impl SignerRunLoop for RunLoop<Signer, T>>::run_one_pass -> Option<Vec<SignerResult>> with Some(vec![Default::default()]) in 0.9s build
3 mutants tested in 2m 29s: 2 missed, 1 unviable
Running mutants for stacks-node (       1 mutants)
Found 1 mutant to test
ok       Unmutated baseline in 147.6s build + 161.3s test
 INFO Auto-set test timeout to 4m 2s
MISSED   testnet/stacks-node/src/config.rs:1352:9: replace Config::get_poll_time -> u64 with 0 in 18.1s build + 161.2s test
1 mutant tested in 8m 9s: 1 missed
Running mutants for stackslib (       1 mutants)
Found 1 mutant to test
ok       Unmutated baseline in 157.7s build + 2054.0s test
 INFO Auto-set test timeout to 51m 21s
caught   stackslib/src/net/chat.rs:730:9: replace ConversationP2P::check_burn_header_hash_disagreement -> bool with true in 39.0s build + 1079.6s test
1 mutant tested in 55m 31s: 1 caught
```

I think the output is a bit too dependent on what packages it runs in order to make the best modular condition to catch those. And I think it is pretty explicit as if there is an error it is displayed, and also if missed or unviable are generated it shows. 

I would assume just making sure all mutants are caught is the result wanted. I would let it with the default return.

Reply:
This is an output example from when it fails to build
```
error: command `/Users/asuciu/.rustup/toolchains/stable-aarch64-apple-darwin/bin/cargo test --no-run --message-format json-render-diagnostics` exited with code 101

*** result: Failure(101)

ERROR cargo build failed in an unmutated tree, so no mutants were tested
Running mutants for stacks-node (       1 mutants)
Found 1 mutant to test
```